File: modules/desktop/home/hyprland/eww/scripts/eww_bar_backend/notifications.py
import subprocess
import threading
import time

from .common import parse_json, run_text, truncate_text

NOTIFICATIONS_DEFAULT = {"paused": "false", "new": 0, "count": 0, "groups": []}

APP_MAX = 20
SUMMARY_MAX = 48
BODY_MAX = 64

# Ephemeral UI state, process-lifetime only (same idea as the display-mode
# file, but collapse/badge state is fine to lose on daemon restart).
_UI_LOCK = threading.Lock()
_COLLAPSED = set()
_LAST_SEEN_US = 0


def _field(item, name, default=""):
    value = item.get(name)
    if isinstance(value, dict):
        return value.get("data", default)
    return default


def parse_history_items(history_json):
    # `dunstctl history` wraps everything in {"type": "aa{sv}", "data": [[...]]}
    # and every field in {"type": ..., "data": ...}. timestamp is MICROSECONDS
    # on the monotonic boot clock, not wall time.
    body = parse_json(history_json, {})
    if not isinstance(body, dict):
        return []
    data = body.get("data")
    if not (isinstance(data, list) and data and isinstance(data[0], list)):
        return []
    items = []
    for raw in data[0]:
        if not isinstance(raw, dict):
            continue
        try:
            item_id = int(_field(raw, "id", 0))
            timestamp = int(_field(raw, "timestamp", 0))
        except (TypeError, ValueError):
            continue
        items.append(
            {
                "id": item_id,
                "app": str(_field(raw, "appname") or "unknown"),
                "summary": str(_field(raw, "summary") or ""),
                "body": str(_field(raw, "body") or ""),
                "urgency": str(_field(raw, "urgency") or "NORMAL"),
                "timestamp": timestamp,
            }
        )
    items.sort(key=lambda item: -item["timestamp"])
    return items


def format_age(seconds):
    if seconds < 10:
        return "now"
    if seconds < 60:
        return f"{int(seconds)}s"
    if seconds < 3600:
        return f"{int(seconds // 60)}m"
    if seconds < 86400:
        return f"{int(seconds // 3600)}h"
    return f"{int(seconds // 86400)}d"


def boottime_seconds():
    # dunst stamps history with CLOCK_BOOTTIME (suspend-included) — verified on
    # this host: timestamps run ahead of CLOCK_MONOTONIC by exactly the
    # accumulated suspend time. Using MONOTONIC here makes ages go negative
    # (clamped to "now") after any suspend.
    try:
        return time.clock_gettime(time.CLOCK_BOOTTIME)
    except Exception:
        try:
            return time.clock_gettime(time.CLOCK_MONOTONIC)
        except Exception:
            return 0.0


def notifications_state_from_parts(items, paused_text, now_monotonic_us, collapsed, last_seen_us):
    grouped = {}
    order = []
    for item in items:
        app = truncate_text(item["app"], APP_MAX)
        if app not in grouped:
            grouped[app] = []
            order.append(app)
        age = max(0.0, (now_monotonic_us - item["timestamp"]) / 1_000_000)
        grouped[app].append(
            {
                "id": item["id"],
                "summary": truncate_text(item["summary"], SUMMARY_MAX),
                "body": truncate_text(item["body"].replace("\n", " "), BODY_MAX),
                "age": format_age(age),
                "urgency": item["urgency"],
            }
        )
    groups = [
        {
            "app": app,
            "count": len(grouped[app]),
            "collapsed": "true" if app in collapsed else "false",
            "items": grouped[app],
        }
        for app in order
    ]
    return {
        "paused": "true" if paused_text.strip() == "true" else "false",
        "new": sum(1 for item in items if item["timestamp"] > last_seen_us),
        "count": len(items),
        "groups": groups,
    }


def notifications_state():
    items = parse_history_items(run_text(["dunstctl", "history"]))
    paused_text = run_text(["dunstctl", "is-paused"])
    with _UI_LOCK:
        collapsed = set(_COLLAPSED)
        last_seen = _LAST_SEEN_US
    return notifications_state_from_parts(
        items, paused_text, boottime_seconds() * 1_000_000, collapsed, last_seen
    )


def _run_dunstctl(args):
    subprocess.run(
        ["dunstctl", *args],
        stdin=subprocess.DEVNULL, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL,
        check=False, timeout=5,
    )


def toggle_group(app):
    if not app:
        raise ValueError("notif toggle-group requires an app name")
    with _UI_LOCK:
        if app in _COLLAPSED:
            _COLLAPSED.discard(app)
        else:
            _COLLAPSED.add(app)
    return notifications_state()


def dismiss_notification(notif_id):
    try:
        value = int(notif_id)
    except (TypeError, ValueError):
        raise ValueError("notif dismiss requires a numeric id")
    _run_dunstctl(["history-rm", str(value)])
    return notifications_state()


def clear_group(app):
    if not app:
        raise ValueError("notif clear-group requires an app name")
    # Group names are the truncated app names, so match with the same truncation.
    for item in parse_history_items(run_text(["dunstctl", "history"])):
        if truncate_text(item["app"], APP_MAX) == app:
            _run_dunstctl(["history-rm", str(item["id"])])
    return notifications_state()


def clear_all_notifications():
    _run_dunstctl(["history-clear"])
    return notifications_state()


def toggle_dnd():
    _run_dunstctl(["set-paused", "toggle"])
    return notifications_state()


def mark_seen():
    global _LAST_SEEN_US
    items = parse_history_items(run_text(["dunstctl", "history"]))
    newest = items[0]["timestamp"] if items else int(boottime_seconds() * 1_000_000)
    with _UI_LOCK:
        _LAST_SEEN_US = max(_LAST_SEEN_US, newest)
    return notifications_state()

File: modules/desktop/home/hyprland/eww/go/internal/collect/notifications.go
package collect

import (
	"bytes"
	"encoding/json"
	"fmt"
	"math"
	"sort"
	"strconv"
	"strings"
)

const (
	appMax     = 20
	summaryMax = 48
	bodyMax    = 64
)

// HistoryItem is one parsed entry of `dunstctl history`.
type HistoryItem struct {
	ID        int    `json:"id"`
	App       string `json:"app"`
	Summary   string `json:"summary"`
	Body      string `json:"body"`
	Urgency   string `json:"urgency"`
	Timestamp int64  `json:"timestamp"`
}

// NotificationItem is one entry as the popup renders it.
type NotificationItem struct {
	ID      int    `json:"id"`
	Summary string `json:"summary"`
	Body    string `json:"body"`
	Age     string `json:"age"`
	Urgency string `json:"urgency"`
}

// NotificationGroup is one app's notifications.
type NotificationGroup struct {
	App       string             `json:"app"`
	Count     int                `json:"count"`
	Collapsed string             `json:"collapsed"`
	Items     []NotificationItem `json:"items"`
}

// NotificationsState is the shape notifications_state_from_parts returns.
type NotificationsState struct {
	Paused string              `json:"paused"`
	New    int                 `json:"new"`
	Count  int                 `json:"count"`
	Groups []NotificationGroup `json:"groups"`
}

// pyStr renders a decoded JSON value the way Python's str() would. The distinction
// Go loses by default is int versus float: str() of 5 and 5.0 is "5" and "5.0", so
// the history is decoded with UseNumber and the original literal kept.
func pyStr(value any) string {
	switch typed := value.(type) {
	case nil:
		return "None"
	case string:
		return typed
	case bool:
		if typed {
			return "True"
		}
		return "False"
	case json.Number:
		return typed.String()
	default:
		return fmt.Sprint(typed)
	}
}

// pyTruthy is Python's truthiness for the values json.Unmarshal produces.
func pyTruthy(value any) bool {
	switch typed := value.(type) {
	case nil:
		return false
	case string:
		return typed != ""
	case bool:
		return typed
	case json.Number:
		if f, err := typed.Float64(); err == nil {
			return f != 0
		}
		return typed.String() != ""
	case float64:
		// Reachable whenever a caller decoded without UseNumber. Without this case
		// the switch fell through to `return true`, so a JSON 0 read as truthy --
		// which made SplitMonitors treat `"disabled": 0` as disabled.
		return typed != 0
	case int:
		return typed != 0
	case int64:
		return typed != 0
	case []any:
		return len(typed) > 0
	case map[string]any:
		return len(typed) > 0
	}
	return true
}

// Note what dunstField does NOT do: if the value is not a {"type":..,"data":..}
// wrapper it returns the default, not the value, so a bare `"appname": "kitty"`
// reads as absent. dunst always wraps.
func dunstField(item map[string]any, name string, def any) any {
	value, present := item[name]
	if !present {
		return def
	}
	wrapper, isWrapper := value.(map[string]any)
	if !isWrapper {
		return def
	}
	inner, present := wrapper["data"]
	if !present {
		return def
	}
	return inner
}

// toInt is Python's int() over the values that reach it: truncates floats toward
// zero, parses integral strings, and reports failure where Python would raise.
func toInt(value any) (int64, bool) {
	switch typed := value.(type) {
	case json.Number:
		if i, err := typed.Int64(); err == nil {
			return i, true
		}
		f, err := typed.Float64()
		if err != nil {
			return 0, false
		}
		return int64(math.Trunc(f)), true
	case string:
		i, err := strconv.ParseInt(strings.TrimSpace(typed), 10, 64)
		if err != nil {
			return 0, false
		}
		return i, true
	case bool:
		if typed {
			return 1, true
		}
		return 0, true
	}
	return 0, false
}

// ParseHistoryItems: `dunstctl history` wraps everything in
// {"type": "aa{sv}", "data": [[...]]} and every field in {"type":.., "data":..}.
// timestamp is MICROSECONDS on the boot clock, not wall time.
func ParseHistoryItems(historyJSON string) []HistoryItem {
	decoder := json.NewDecoder(bytes.NewReader([]byte(historyJSON)))
	decoder.UseNumber() // keep 5 distinct from 5.0, see pyStr
	var body map[string]any
	if err := decoder.Decode(&body); err != nil {
		return []HistoryItem{}
	}

	data, _ := body["data"].([]any)
	if len(data) == 0 {
		return []HistoryItem{}
	}
	first, ok := data[0].([]any)
	if !ok {
		return []HistoryItem{}
	}

	items := []HistoryItem{}
	for _, raw := range first {
		entry, ok := raw.(map[string]any)
		if !ok {
			continue
		}
		id, ok := toInt(dunstField(entry, "id", json.Number("0")))
		if !ok {
			continue
		}
		timestamp, ok := toInt(dunstField(entry, "timestamp", json.Number("0")))
		if !ok {
			continue
		}
		items = append(items, HistoryItem{
			ID:        int(id),
			App:       orDefault(dunstField(entry, "appname", ""), "unknown"),
			Summary:   orDefault(dunstField(entry, "summary", ""), ""),
			Body:      orDefault(dunstField(entry, "body", ""), ""),
			Urgency:   orDefault(dunstField(entry, "urgency", ""), "NORMAL"),
			Timestamp: timestamp,
		})
	}

	// SliceStable, not Slice: dunst can stamp two notifications with the same
	// microsecond, and an unstable sort would reorder them between runs, which the
	// emit loop would see as a change and push to eww for no reason.
	sort.SliceStable(items, func(a, b int) bool {
		return items[a].Timestamp > items[b].Timestamp
	})
	return items
}

// orDefault is Python's `str(x or fallback)`.
func orDefault(value any, fallback string) string {
	if !pyTruthy(value) {
		return fallback
	}
	return pyStr(value)
}

func FormatAge(seconds float64) string {
	switch {
	case seconds < 10:
		return "now"
	case seconds < 60:
		return fmt.Sprintf("%ds", int64(seconds))
	case seconds < 3600:
		return fmt.Sprintf("%dm", int64(math.Floor(seconds/60)))
	case seconds < 86400:
		return fmt.Sprintf("%dh", int64(math.Floor(seconds/3600)))
	default:
		return fmt.Sprintf("%dd", int64(math.Floor(seconds/86400)))
	}
}

func NotificationsStateFromParts(
	items []HistoryItem,
	pausedText string,
	nowMonotonicUS float64,
	collapsed map[string]bool,
	lastSeenUS int64,
) NotificationsState {
	// grouped + order, not a plain map: Go map iteration is randomised, and
	// first-appearance order is what puts the most recent app at the top.
	grouped := map[string][]NotificationItem{}
	var order []string

	for _, item := range items {
		app := TruncateText(item.App, appMax)
		if _, seen := grouped[app]; !seen {
			grouped[app] = []NotificationItem{}
			order = append(order, app)
		}
		// NOT covered by the equivalence gate and cannot be: FormatAge returns "now"
		// for everything below 10. Kept because a negative age is reachable -- dunst
		// stamps CLOCK_BOOTTIME, and a notification can be newer than the caller's
		// reading taken a moment earlier.
		age := math.Max(0, (nowMonotonicUS-float64(item.Timestamp))/1_000_000)
		grouped[app] = append(grouped[app], NotificationItem{
			ID:      item.ID,
			Summary: TruncateText(item.Summary, summaryMax),
			Body:    TruncateText(strings.ReplaceAll(item.Body, "\n", " "), bodyMax),
			Age:     FormatAge(age),
			Urgency: item.Urgency,
		})
	}

	groups := []NotificationGroup{}
	for _, app := range order {
		collapsedFlag := "false"
		if collapsed[app] {
			collapsedFlag = "true"
		}
		groups = append(groups, NotificationGroup{
			App:       app,
			Count:     len(grouped[app]),
			Collapsed: collapsedFlag,
			Items:     grouped[app],
		})
	}

	newCount := 0
	for _, item := range items {
		if item.Timestamp > lastSeenUS {
			newCount++
		}
	}

	paused := "false"
	if Strip(pausedText) == "true" {
		paused = "true"
	}
	return NotificationsState{Paused: paused, New: newCount, Count: len(items), Groups: groups}
}
